### ISSUE TITLE
Gogs?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/github-backup#136]
Our original reasoning for this project is still very much still a reality. 

GitHub still has regular "incidents": https://www.githubstatus.com/history where it's offline for 
![image](https://user-images.githubusercontent.com/194400/162309704-ced0641a-8ee3-4e8b-bc40-89d1c56a5a3b.png)

Using GitHub as our [Single source of truth](https://en.wikipedia.org/wiki/Single_source_of_truth) means we need access to it

At present this project mirrors issue data to **`AWS S3`** and @RobStallion asked a very valid question about it: #126 
That's _OK_ as a "cold" _backup_ but it's not very useful for business continuity.

## Starting Point

+ [ ] Site: https://gogs.io/
+ [ ] Installation: https://gogs.io/docs/installation
+ [ ] Deploy to Fly.io: https://fly.io/docs/app-guides/git-gogs-server/

# Todo

+ [ ] Try the demo: https://try.gogs.io
+ [ ] Install Gogs on a test Mac
  + [ ] Play around with it and see what features it has
+ [ ] Deploy to Fly.io with lowest possible resources
+ [ ] Figure out how to secure it. (read docs!)
+ [ ] Build an interface to connect to the `Gogs` server from `Elixir`: https://github.com/dwyl/gogs
+ [ ] Backup GitHub issues for a sample repo (e.g. this one) to the `Gogs` server on Fly.io

**Comment by nelsonic** [comment]
https://github.com/gogs/gogs GitHub repo is very popular and has a healthy number of contributors:
![image](https://user-images.githubusercontent.com/194400/162426939-fba18359-ec62-4d18-b258-8443b02360d0.png)

Who is using `Gogs`? https://gogs.io a lot of CN companies:
![image](https://user-images.githubusercontent.com/194400/162427108-af3abc8c-b38a-4477-8fad-3faf0c7f9aee.png)

Which makes sense as GitHub is partially blocked in China: https://en.wikipedia.org/wiki/Censorship_of_GitHub#China
and it probably makes sense for CN companies to be at least a little paranoid of putting all their code/secrets on a US-owned platform. 💭

**Comment by nelsonic** [comment]
https://news.ycombinator.com/item?id=31033758

**Comment by nelsonic** [comment]
**`Gogs`** server deployed to: https://gogs-server.fly.dev/ 🚀 
![gogs-server](https://user-images.githubusercontent.com/194400/165000758-0ca9e54d-2c8a-429f-9c84-ff4bfd68bed2.png)

Documented in: https://github.com/dwyl/gogs-server ✅

**Comment by nelsonic** [comment]
The work on https://github.com/dwyl/gogs is coming along nicely. 
We've shipped _most_ of the core functions we need: https://hexdocs.pm/gogs/api-reference.html

But we are tidying it up a bit first before _using_ it in an App: https://github.com/dwyl/gogs/issues/24 + https://github.com/dwyl/gogs/pull/25 + https://github.com/dwyl/gogs/issues/26

**Comment by nelsonic** [comment]
_really_ wish I had seen this page earlier: https://docs.gitea.io/en-us/comparison/ ... 🙄  ⏳ 
![image](https://user-images.githubusercontent.com/194400/168423015-05980f2c-1a81-4369-a8f4-2a118bedf847.png)

**Comment by nelsonic** [comment]
We've invested a _considerable_ amount of time in: ⌛ 
+ Deploying `Gogs`: https://github.com/dwyl/gogs-server + https://github.com/dwyl/learn-devops/tree/main/gogs 🚀 
+ Creating a package to interface with a `Gogs` instance from `Elixir` >`gogs`: https://github.com/dwyl/gogs ⚙️ 
+ and a demo app showcasing that interface: https://github.com/dwyl/gogs-demo ✨ 

But re-reading the `Gitea` API I'm being forced to re-consider ... 🙄 
I'm suffering from a _serious_ case of ["**Sunk Cost _Bias_**"](https://en.wikipedia.org/wiki/Sunk_cost#Fallacy_effect) ...

Basically **3 things** are forcing me consider switching from **`Gogs`** to ***`Gitea`***:

1. More extensive and documented API: https://try.gitea.io/api/swagger#/
2. Ability to create Orgs via the API: 
![image](https://user-images.githubusercontent.com/194400/168423884-6f65525b-927c-4b8f-bd5c-a9c7cb4ea8df.png)

  + Orgs have a **`visibility`** option which allows them to be _unlisted_ 
  i.e. not visible even if someone has access to the Gitea instance.

**Comment by nelsonic** [comment]
Ok, `while` it **_totally_ sucks** to write-off the time we've invested in `Gogs` and switch to `Gitea`, ⏳ 🔥 🤦‍♂️ 
I'm thinking that I will do a **`Spike`** to try and replicate the work we've done (deployment, module & demo). 🧑‍💻 
It's _definitely_ not a _complete_ waste of effort because all the ***`local_...`*** functions can be recycled ♻️ 
and my reading is that there is a migration path from `Gogs` to `Gitea`: https://docs.gitea.io/en-us/upgrade-from-gogs ...
So all the underlying "DevOps" work should be transferable too. 🤞 

Going to start with getting `Gitea` running on `Fly.io`: https://github.com/dwyl/gitea-server
Then I will try and simulate the API requests noted above:
1. Create an org that has `visibility: hidden` 
2. Create a PR and attempt to get the `diff` via the REST API

If those two things work and it doesn't take me too much time I will report back.

Wish me luck! 🍀

**Comment by nelsonic** [comment]
Not for `github-backup`, but this is a _useful_ feature for orgs that need privacy: 
https://gitea-server.fly.dev/private-org visible to members:
![image](https://user-images.githubusercontent.com/194400/168494467-3e116e6f-262c-4c76-939d-f822d099f8a7.png)

404 for anyone else:
![image](https://user-images.githubusercontent.com/194400/168494489-69294fa1-519c-405b-9654-22c6932691ce.png)

**Comment by nelsonic** [comment]
We've made substantial progress with this:
+ `gitea`: https://github.com/dwyl/gitea
+ `gitea-server`: https://github.com/dwyl/gitea-server
+ `gitea-demo`: https://github.com/dwyl/gitea-demo

Still `#Todo`: Backups! https://github.com/dwyl/gitea-server/issues/3 🙏
